gamemaster/adjudicator/pregame.php

```diff
@@ -1,322 +1,231 @@
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU Affero General Public License
+    along with webDiplomacy.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+defined('IN_CODE') or die('This script can not be run by itself.');
+
+require_once('assignmentSolver.php');
+
+/**
+ * @package GameMaster
+ * @subpackage Adjudicator
+ */
+class adjudicatorPreGame {
+
+	protected function isEnoughPlayers() {
+		global $Game;
+
+		return ( count($Game->Members->ByID) == count($Game->Variant->countries) );
+	}
+
+	protected function userCountries() {
+		global $Game;
+		/*
+		 * Find out who gets which countryID;
+		 * - Get the number of times each player has played as each countryID (stored in wD_Users)
+		 * - Assign players to countries so that they play as a country that they have played
+		 *   as as few times as possible.
+		 */
+
+		$userIDs = array();
+		foreach($Game->Members->ByID as $M) $userIDs[]=$M->userID;
+		
+		/**
+		 * The hungarian solver expects a multidimensional array, all 0-indexed. Therefore
+		 * we need to store the userIDs seperately to the weights, so that they can be 
+		 * zipped together at the end.
+		 */
+		$weights = $this->getWeightsAsCountryCounts($userIDs);
+
+		$solver = new assignmentSolver;
+		
+		// We are minimising the weights, so we pass in "false".
+		$assignment = $solver->hungarian($weights, false);
+		
+		$userCountries = array();
+		/*
+		 * Here translate from 0-index to 1-index, and zipping together the userIDs and 
+		 * the assignments.
+		 */
+		for($i = 0; $i < count($userIDs); $i++)
+		{
+			$userID = $userIDs[$i];
+			$userCountries[$userID] = $assignment[$i] + 1;
+		}
+			
+		/**
+		 * Update the play count for the associated countries
+		 */
+		$this->updateCountryPlayCounts($assignment, $userIDs, $weights);			
+
+		return $userCountries;
+	}
+	
+	protected function updateCountryPlayCounts($assignment, $userIDs, $initialWeights)
+	{
+		global $Game;
+		$vd = new VariantData($Game->variantID);
+		$vd->systemToken = 948379409;
+		for ($userIndex = 0; $userIndex < count($userIDs); $userIndex++)
+		{
+			$vd->userID = $userIDs[$userIndex];
+			$countryIndex = $assignment[$userIndex];
+			$oldPlayCount = $initialWeights[$userIndex][$countryIndex];
+			$vd->updateInt($oldPlayCount + 1, $countryIndex + 1);
+		}
+	}
+	
+	protected function getWeightsAsCountryCounts($userIDs)
+	{
+		global $Game;
+		$vd = new VariantData($Game->variantID);
+		$vd->systemToken = 948379409;
+		
+		$userCountryCounts = array();
+		
+		$countryCount = count($Game->Variant->countries);
+		for($i = 0; $i < count($userIDs); $i++)
+		{
+			$userCountryCounts[$i] = array();
+			$vd->userID = $userIDs[$i];
+			
+			/**
+			 * Note that the first country ID is 1 not 0.
+			 * assignmentSolver requires that the indexes start at 0, so we need to 
```

Comment:
It feels like `assignmentSolver` should do this conversion, rather than expecting it to be done before the call. This way the contract for `assignmentSolver` is written in the same data schema as the rest of the code.

Reply:
Is it possible that countryId might one day not start at 1? Or some weird variant might have a particular way of assigning country IDs? That's why I had the conversion here, to keep `assignmentSolver` as general as possible. I will still make the change if it's best.

Reply:
I like the idea of being robust to future changes - and I think that's a good argument for keeping the same 1-based schema right up to the solver boundary (converting inside the solver). Any changes to what a `countryId` means would be complicated by multiple definitions.

(Also, the code has the assumption of starting from 1, with 0 meaning "everyone" in many many places, so I don't think it'll change)